hydro_cli_maelstrom/src/main.rs

```diff
@@ -0,0 +1,351 @@
+
+    let mut lines = FramedRead::new(in_stream, LengthDelimitedCodec::new());
+    #[cfg(debug_assertions)]
+    println!("accepted connection");
+
+    // Itializer counter which tracks the next available msg_id
+    let mut msg_id_counter = output_id;
+
+    while let Some(Ok(line)) = lines.next().await {
+        // Transforms output into maelstrom payload
+        // For example:
+        // {"echo":"hello world!","in_reply_to":["n1", 1]}
+        // ->
+        // {"src":"n1","dest":"c1","body":{"echo":"hello world!","msg_id":0,"in_reply_to":1,"type":"echo_ok"}}
+
+        // parse line to string
+        let raw_line: String = bincode::deserialize(&line).unwrap();
+
+        // parse raw string to json value
+        let mut v: Value = serde_json::from_str(&raw_line)?;
+        let body = v.as_object_mut().unwrap();
+        body.insert("type".to_string(), maelstrom_type.clone().into());
+
+        // in_reply_to is actually [src, msg_id]
+        let in_reply_to = v["in_reply_to"].as_array_mut().unwrap();
+        assert_eq!(
+            in_reply_to.len(),
+            2,
+            "in_reply_to was not a pair of [src, msg_id]"
+        );
+        let dest = in_reply_to.remove(0);
+        let in_reply_to = in_reply_to.remove(0);
+
+        // set in_reply_to to real value
+        v["in_reply_to"] = in_reply_to;
+
+        // set msg_id to next available message id
+        v["msg_id"] = Value::Number(msg_id_counter.into());
+        msg_id_counter += output_count;
+
+        // wrap v in maelstrom header
+        let response = json!({
+            "src": node_id.clone(),
+            "dest": dest,
+            "body": v
+        });
+
+        // send it to stdout
+        println!("{}", response);
+    }
+    Ok(())
+}
+
+#[tokio::main]
+async fn main() -> Result<()> {
+    // let path = r"C:\Users\rhala\Code\hydroflow\target\debug\examples\echo";
+    let path = r"/mnt/c/Users/rhala/Code/hydroflow/target/debug/examples/maelstrom_unique_id";
+    let mut child = Command::new(path)
+        .stdin(Stdio::piped())
+        .stdout(Stdio::piped())
+        .stderr(Stdio::piped())
+        .spawn()?;
+
+    let mut child_stdin = child.stdin.take().unwrap();
+    let mut child_stdout = child.stdout.take().unwrap();
+    let child_stderr = child.stderr.take().unwrap();
+
+    // this is supposed to be an input to the program
+    // vec![
+    // PortDefn::new("echo".into(), "echo_in".into(), Direction::Source),
+    // PortDefn::new("echo_ok".into(), "echo_out".into(), Direction::Sink),
+    // ];
+    // echo setup
+    // let port_input_str = r#"[{"maelstrom_type":"echo","port_name":"echo_in","direction":"Source"},{"maelstrom_type":"echo_ok","port_name":"echo_out","direction":"Sink"}]"#;
+
+    // unique id gen setup
+    let port_input_str = r#"[{"maelstrom_type":"generate","port_name":"gen_in","direction":"Source"},{"maelstrom_type":"generate_ok","port_name":"ok_out","direction":"Sink"}]"#;
```

Comment:
We should declare Rust datatypes for each of these messages and use `serde_json` to serialize them.